### PR TITLE
[Bug] Add Non-ECS Checks to New Terms Rule Validation

### PR DESCRIPTION
**Pull request by terrancedejesus**
# Related PRs
* https://github.com/elastic/detection-rules/pull/2434
* https://github.com/elastic/detection-rules/pull/2433

## Summary
The referenced pull requests were created to add new detection rules leveraging New Terms rule types, a feature released in  Elastic Stack version 8.4. The current validation had a bug and needed additional logic to check the `non-ecs.json` file for valid fields.

## Additional Information

* Bug -  The `stack_version` variable was storing the value of `feature_min_stack_version` which is a statically defined value of `8.4`. The `stack_version` should instead be the `min_stack_version` defined in the rule.
* Enhanced Validation - In addition to ECS schema validation, we needed to load the `non-ecs.json` file for extra field existence checks. If the field value defined in the rule is not in ECS schema, we should check Non-ECS schema's as well. To do so, the index patterns of the rule are used as keys to the Non-ECS data which returns a JSON object of `field:fieldtype`.

The following example rules were used locally during testing to ensure validation logic worked as expected with these new changes.

<details>
<summary>FirstTime Seen Account Performing DCSync</summary>

```sql
[metadata]
creation_date = "2022/12/19"
maturity = "production"
min_stack_comments = "New fields added: required_fields, related_integrations, setup, New Term rule type"
min_stack_version = "8.4.0"
updated_date = "2022/12/19"

[rule]
author = ["Elastic"]
description = """
This rule identifies when a User Account starts the Active Directory Replication Process for the first time. 
Attackers can use the DCSync technique to get credential information of individual accounts or the entire domain, 
thus compromising the entire domain.
"""
from = "now-9m"
index = ["winlogbeat-*", "logs-system.*"]
language = "kuery"
license = "Elastic License v2"
name = "FirstTime Seen Account Performing DCSync"
note = """## Triage and analysis
### Investigating Potential Credential Access via DCSync
Active Directory replication is the process by which the changes that originate on one domain controller are
automatically transferred to other domain controllers that store the same data.
Active Directory data consists of objects that have properties, or attributes. Each object is an instance of an object
class, and object classes and their respective attributes are defined in the Active Directory schema. Objects are
defined by the values of their attributes, and changes to attribute values must be transferred from the domain
controller on which they occur to every other domain controller that stores a replica of an affected object.
Adversaries can use the DCSync technique that uses Windows Domain Controller's API to simulate the replication process
from a remote domain controller, compromising major credential material such as the Kerberos krbtgt keys used
legitimately for tickets creation, but also tickets forging by attackers. This attack requires some extended privileges
to succeed (DS-Replication-Get-Changes and DS-Replication-Get-Changes-All), which are granted by default to members of
the Administrators, Domain Admins, Enterprise Admins, and Domain Controllers groups. Privileged accounts can be abused
to grant controlled objects the right to DCsync/Replicate.
More details can be found on [Threat Hunter Playbook](https://threathunterplaybook.com/library/windows/active_directory_replication.html?highlight=dcsync#directory-replication-services-auditing) and [The Hacker Recipes](https://www.thehacker.recipes/ad/movement/credentials/dumping/dcsync).
This rule monitors for Event ID 4662 (Operation was performed on an Active Directory object) and identifies events that
use the access mask 0x100 (Control Access) and properties that contain at least one of the following or their equivalent:
Schema-Id-GUID (DS-Replication-Get-Changes, DS-Replication-Get-Changes-All, DS-Replication-Get-Changes-In-Filtered-Set).
It also filters out events that use computer accounts and also Azure AD Connect MSOL accounts (more details [here](https://techcommunity.microsoft.com/t5/microsoft-defender-for-identity/ad-connect-msol-user-suspected-dcsync-attack/m-p/788028)).
#### Possible investigation steps
- Identify the user account that performed the action and whether it should perform this kind of action.
- Contact the account and system owners and confirm whether they are aware of this activity.
- Investigate other alerts associated with the user/host during the past 48 hours.
- Correlate security events 4662 and 4624 (Logon Type 3) by their Logon ID (`winlog.logon.id`) on the Domain Controller
(DC) that received the replication request. This will tell you where the AD replication request came from, and if it
came from another DC or not.
- Scope which credentials were compromised (for example, whether all accounts were replicated or specific ones).
### False positive analysis
- This activity should not happen legitimately, since replication should be done by Domain Controllers only. Any
potential benign true positive (B-TP) should be mapped and monitored by the security team. Any account that performs
this activity can put the domain at risk for not having the same security standards as computer accounts (which have
long, complex, random passwords that change frequently), exposing it to credential cracking attacks (Kerberoasting,
brute force, etc.).
### Response and remediation
- Initiate the incident response process based on the outcome of the triage.
- If specific credentials were compromised:
  - Reset the password for these accounts and other potentially compromised credentials, like email, business systems,
  and web services.
- If the entire domain or the `krbtgt` user were compromised:
  - Activate your incident response plan for total Active Directory compromise which should include, but not be limited
  to, a password reset (twice) of the `krbtgt` user.
- Investigate how the attacker escalated privileges and identify systems they used to conduct lateral movement. Use this
information to scope ways that the attacker could use to regain access to the environment.
- Determine the initial vector abused by the attacker and take action to prevent reinfection through the same vector.
- Using the incident response data, update logging and audit policies to improve the mean time to detect (MTTD) and the
mean time to respond (MTTR).
## Setup
The 'Audit Directory Service Changes' logging policy must be configured for (Success, Failure).
Steps to implement the logging policy with Advanced Audit Configuration:
```
Computer Configuration >
Policies >
Windows Settings >
Security Settings >
Advanced Audit Policies Configuration >
Audit Policies >
DS Access >
Audit Directory Service Changes (Success,Failure)
```
If enabling an EQL rule on a non-elastic-agent index (such as beats) for versions <8.2, events will not define `event.ingested` and default fallback for EQL rules was not added until 8.2, so you will need to add a custom pipeline to populate `event.ingested` to @timestamp for this rule to work.
"""
references = [
    "https://threathunterplaybook.com/notebooks/windows/06_credential_access/WIN-180815210510.html",
    "https://threathunterplaybook.com/library/windows/active_directory_replication.html?highlight=dcsync#directory-replication-services-auditing",
    "https://github.com/SigmaHQ/sigma/blob/master/rules/windows/builtin/security/win_ad_replication_non_machine_account.yml",
    "https://github.com/atc-project/atomic-threat-coverage/blob/master/Atomic_Threat_Coverage/Logging_Policies/LP_0027_windows_audit_directory_service_access.md",
    "https://attack.stealthbits.com/privilege-escalation-using-mimikatz-dcsync",
    "https://www.thehacker.recipes/ad/movement/credentials/dumping/dcsync",
]
risk_score = 73
rule_id = "5c6f4c58-b381-452a-8976-f1b1c6aa0def"
severity = "high"
tags = ["Elastic", "Host", "Windows", "Threat Detection", "Credential Access", "Active Directory", "has_guide"]
timestamp_override = "event.ingested"
type = "new_terms"

query = '''
event.action : "Directory Service Access" and event.code : "4662" and 
 winlog.event_data.Properties : (*DS-Replication-Get-Changes* or *DS-Replication-Get-Changes-All* or *DS-Replication-Get-Changes-In-Filtered-Set* or *1131f6ad-9c07-11d1-f79f-00c04fc2dcd2* or *1131f6aa-9c07-11d1-f79f-00c04fc2dcd2* or *89e95b76-444d-4c62-991a-0facbeda640c*) and 
 not winlog.event_data.SubjectUserName : (*$ or MSOL_*)
'''


[[rule.threat]]
framework = "MITRE ATT&CK"

[[rule.threat.technique]]
id = "T1003"
reference = "https://attack.mitre.org/techniques/T1003/"
name = "OS Credential Dumping"

    [[rule.threat.technique.subtechnique]]
    id = "T1003.006"
    reference = "https://attack.mitre.org/techniques/T1003/006/"
    name = "DCSync"


[rule.threat.tactic]
id = "TA0006"
reference = "https://attack.mitre.org/tactics/TA0006/"
name = "Credential Access"


[rule.new_terms]
field = "new_terms_fields"
value = ["winlog.event_data.SubjectUserName"]

[[rule.new_terms.history_window_start]]
field = "history_window_start"
value = "now-15d"

```

</details>

<details>
<summary>First Time Seen Driver Loaded</summary>

```sql
[metadata]
creation_date = "2022/12/19"
maturity = "production"
min_stack_comments = "New fields added: required_fields, related_integrations, setup, New Term"
min_stack_version = "8.6.0"
updated_date = "2022/12/19"

[rule]
author = ["Elastic"]
description = """
Identifies the load of a driver with an original file name and signature values that were observed 
for the first time during the last 30 days. This rule type can help baseline drivers installation 
within your environment.
"""
from = "now-9m"
index = ["logs-endpoint.events.*"]
language = "kuery"
license = "Elastic License v2"
name = "First Time Seen Driver Loaded"
references = ["https://www.elastic.co/kr/security-labs/stopping-vulnerable-driver-attacks"]
risk_score = 47
rule_id = "df0fd41e-5590-4965-ad5e-cd079ec22fa9"
severity = "medium"
tags = ["Elastic", "Host", "Windows", "Threat Detection", "Persistence"]
timestamp_override = "event.ingested"
type = "new_terms"

query = '''
event.category : "driver" and event.action : "load" 
'''


[[rule.threat]]
framework = "MITRE ATT&CK"
[[rule.threat.technique]]
id = "T1543"
name = "Create or Modify System Process"
reference = "https://attack.mitre.org/techniques/T1543/"
[[rule.threat.technique.subtechnique]]
id = "T1543.003"
name = "Windows Service"
reference = "https://attack.mitre.org/techniques/T1543/003/"


[rule.threat.tactic]
id = "TA0003"
name = "Persistence"
reference = "https://attack.mitre.org/tactics/TA0003/"

[rule.new_terms]
field = "new_terms_fields"
value = ["dll.pe.original_file_name", "dll.code_signature.subject_name"]
[[rule.new_terms.history_window_start]]
field = "history_window_start"
value = "now-30d"
```

</details>

Since the new rules are in another PR, I have shared the output of pytests and `view-rule` CLI command which goes through the same logic to do validation on the rule before outputting API format.

Screen Shot 2022-12-19 at 10 19 57 AM

## Testing
If anyone would like to test, copy the rules saved above into the Windows rule path. Run `pytest tests` for unit testing or `python -m detection_rules view-rule RULEPATH` to test the `validate` method under `NewTermsRuleData` class.

### CODE REVIEW

File: detection_rules/rule.py [hunk 1]
```diff
@@ -481,9 +481,10 @@ class HistoryWindowStart:
     def validate(self, meta: RuleMeta) -> None:
         """Validates terms in new_terms_fields are valid ECS schema."""
 
-        super(NewTermsRuleData, self).validate_query(meta)
-        feature_min_stack = '8.4.0'
-        feature_min_stack_extended_fields = '8.6.0'
+        kql_validator = KQLValidator(self.query)
```
Comment (Mikaayenson) [inline detection_rules/rule.py:484]:
```suggestion
        assert self.language == 'kquery', \ 
            "new terms field only supports the kquery language" 
        kql_validator = KQLValidator(self.query)
```

**Review by Mikaayenson — APPROVED** [review]

File: detection_rules/rule.py (hunk 1 above)
Comment (terrancedejesus) [reply detection_rules/rule.py:484]: Confirmed `mixins.MarshmallowDataclassMixin.from_dict` will fail with schema validation errors if `language` is eql. Attempted to add that as the rule type as well and if failed too.

Comment (Mikaayenson) [reply detection_rules/rule.py:484]: due to FilterLanguages = Literal["kuery", "lucene"]

**Review by w0rk3r — APPROVED** [review]
LGTM

**Pull request merged**